### ISSUE TITLE
TypeError: can't compare offset-naive and offset-aware datetimes

### ISSUE THREAD

**Issue by LeptoFlare** [scrapinghub/dateparser#679]
This issue is not isolated inside the function. It permeates outside of the function.

It seems to start occurring after you use search_dates on relative and absolute datetimes at the same time.

Just use search_dates on this string:
```
May 2020
June 2020
2023
January UTC
June 5 am utc
June 23th 5 pm EST
May 31, 8am UTC
```

**Comment by LeptoFlare** [comment]
```py
>>> from dateparser.search import search_dates
>>> search_dates("""May 2020
... June 2020
... 2023
... January UTC
... June 5 am utc
... June 23th 5 pm EST
... May 31, 8am UTC""")

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/search/__init__.py", line 49, in search_dates
    result = _search_with_detection.search_dates(
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/conf.py", line 84, in wrapper
    return f(*args, **kwargs)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/search/search.py", line 229, in search_dates
    return {'Language': language_shortname, 'Dates': self.search.search_parse(language_shortname, text,
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/search/search.py", line 161, in search_parse
    parsed, substrings = self.parse_found_objects(parser=parser, to_parse=translated,
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/search/search.py", line 125, in parse_found_objects
    parsed_item = self.parse_item(parser, item, translated[i], parsed, need_relative_base)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/search/search.py", line 111, in parse_item
    parsed_item = parser.get_date_data(item)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/date.py", line 417, in get_date_data
    parsed_date = _DateLocaleParser.parse(
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/date.py", line 196, in parse
    return instance._parse()
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/date.py", line 200, in _parse
    date_obj = self._parsers[parser_name]()
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/date.py", line 222, in _try_parser
    date_obj, period = date_parser.parse(
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/conf.py", line 84, in wrapper
    return f(*args, **kwargs)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/date_parser.py", line 26, in parse
    date_obj, period = parse(date_string, settings=settings)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/parser.py", line 72, in parse
    raise exceptions.pop(-1)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/parser.py", line 66, in parse
    res = parser(datestring, settings)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/parser.py", line 443, in parse
    dateobj = po._correct_for_time_frame(dateobj)
  File "/home/leptospira/lohi/.venv/lib/python3.8/site-packages/dateparser/parser.py", line 395, in _correct_for_time_frame
    if self.now < dateobj:
TypeError: can't compare offset-naive and offset-aware datetimes
```

**Comment by noviluni** [comment]
Related to: https://github.com/scrapinghub/dateparser/issues/491

**Comment by noviluni** [comment]
I've been investigating this. It is a weird bug, as the next works:

```python3
dates = ['May 2020', 'June 2020', '2023', 'January UTC', 'June 5 am utc', 'June 23th 5 pm EST', 'May 31, 8am UTC']

for date in dates:
    print(search_dates(date))
```

But when joining the dates it doesn't work:

```python3
print(search_dates('\n'.join(dates)))
```

This is because of the `self.now` of the `parser.py`. 

I found a workaround: https://github.com/scrapinghub/dateparser/pull/736 but it seems that the core problem is the `self.now`. It's also the causing of some errors related to the multithreading: https://github.com/scrapinghub/dateparser/issues/441

I will check it when having a moment.

**Comment by jhawgs** [comment]
I am working on a project that utilizes this search functionality, and I get the same error. Do you know whether there is a fix right around the corner? Otherwise, I’m happy to work on one. I just don’t want to submit an unnecessary pull if there’s gonna be a fix soon.

**Comment by noviluni** [comment]
Hi @jhawgs, sorry, I haven't time to spend on this. Please, go ahead!

I created this PR: https://github.com/scrapinghub/dateparser/pull/736, you can continue from that or submit a new/different PR. The only requirement is to add tests to check that nothing is broken and the coverage is still the same after merging the PR.

Thanks!

**Comment by jhawgs** [comment]
I've been working on the bug, but I don't believe that I am familiar enough with the code as I would need to be to put together a comprehensive fix. I did, however, find that the error occurs when running a function that only seems to be useful when one has set the PREFER_DATES_FROM setting to something other than the default. Because I am not using this functionality, I am simply bypassing the function call. To do this I comment out the following line in parser.py:

`dateobj = po._correct_for_time_frame(dateobj)`

This may or may not be an option for anybody having the problem, but I certainly do not recommend it as a long term fix.

Also, I found some weird behavior that I don't think is expected, but I might be wrong. When running LeptoFlare's example, the parser tries to retain certain states across dates. This portion of the example:

`>>> search_dates("""May 2020
... June 2020
... 2023
... January UTC""")`

gets the following dates:

`2020-05-14 00:00:00
2020-06-14 00:00:00
2023-01-14 00:00:00
2023-01-14 00:00:00`

Specifically it assumes that "January UTC" is in 2023 instead of reverting back to the current year and assuming that it refers to January of either 2021 or 2020. This might be expected behavior when leaving PREFER_DATES_FROM as the default "current_period," but when looking at the code that should change the date based on a new setting of "past" or "future," it looks like the only possibilities would be 2022 or 2023. This bug occurs in the same function call that I commented out, which suggests that anybody looking for a comprehensive fix for either issue should start at:

`dateobj = po._correct_for_time_frame(dateobj)` in parser.py.

**Comment by SimplicityGuy** [comment]
Here's another string type that has the same error thrown: `[02.10.21]`

Granted this is ambiguous as to whether it's February 10, 2021 or October 2, 2021, however the same `can't compare offset-naive and offset-aware datetimes` is thrown

**Comment by SimplicityGuy** [comment]
The issue here is not one of multi-threading or `self.now`. The issue is that `self.now` gets set to a timezone unaware (e.g. naive) time. I'm investigating a fix here.

**Comment by SimplicityGuy** [comment]
Fixed in PR #881. Please take a look.

**Comment by noviluni** [comment]
PR #881 has been merged, so we can close this issue now :)
